src/NuGet.Clients/NuGet.PackageManagement.UI/Models/Package/PackageModelFactory.cs

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using NuGet.PackageManagement.VisualStudio;
+using NuGet.VisualStudio.Internal.Contracts;
+using ContractItemFilter = NuGet.VisualStudio.Internal.Contracts.ItemFilter;
+
+namespace NuGet.PackageManagement.UI.Models.Package
+{
+    internal class PackageModelFactory
+    {
+        private readonly INuGetSearchService _searchService;
+        private readonly INuGetPackageFileService _packageFileService;
+        private readonly IPackageVulnerabilityService _packageVulnerabilityService;
+        private readonly bool _includePrerelease;
+        private IReadOnlyCollection<PackageSourceContextInfo> _packageSources;
+
+        public PackageModelFactory(INuGetSearchService searchService, INuGetPackageFileService packageFileService, IPackageVulnerabilityService packageVulnerabilityService, bool includePrerelease, IReadOnlyCollection<PackageSourceContextInfo> packageSources)
+        {
+            _searchService = searchService ?? throw new ArgumentNullException(nameof(_searchService));
+            _packageFileService = packageFileService ?? throw new ArgumentNullException(nameof(_packageFileService));
+            _packageVulnerabilityService = packageVulnerabilityService ?? throw new ArgumentNullException(nameof(_packageVulnerabilityService));
+            _includePrerelease = includePrerelease;
+            _packageSources = packageSources ?? throw new ArgumentNullException(nameof(_packageSources));
+        }
+
+        public PackageModel Create(PackageSearchMetadataContextInfo metadata, ContractItemFilter itemFilter)
+        {
+            if (metadata == null)
+            {
+                throw new ArgumentNullException(nameof(metadata));
+            }
+
+            EmbeddedResourcesCapability embeddedResources = new EmbeddedResourcesCapability(_packageFileService, metadata.Identity!, metadata.ReadmeUrl);
+
+            if (metadata.PackagePath != null)
+            {
+                PackageMetadataRetrievalAdapter packageMetadataRetrievalAdapter = new PackageMetadataRetrievalAdapter(_searchService, metadata.Identity!, _packageSources, _includePrerelease);
+                IVulnerableCapable vulnerableCapability = new VulnerablePackageMetadataCapability(packageMetadataRetrievalAdapter);
+
+                if (itemFilter.Equals(ContractItemFilter.All))
+                {
+                    // Package from a local folder
+                    return new LocalPackageModel(
```

Comment:
This method is hard to understand the flow for, since all of the `return new ...` statements are nearly 20 lines long, causing the method to be multiple screens long.

If you create helper methods to create the specific models, so these statements become `return Create???(metadata, other, params);`, then then all the if-else branches should be viewable on a single screen, making it easier to follow.

Reply:
Agreed, thanks! I have now split this into helper methods to make it easier to read.